Protect against null Widgets

andrewkaufman opened the pull request:
We have been getting several errors when loading Gaffer inside Maya (2016.sp6, 2017.u5 and 2018.5). They're fairly innocuous but clutter Maya's ScriptEditor making it hard to debug more serious issues. The particular repro case was:

- Open Maya
- Launch Caribou (Gaffer in Maya)
- Open a Maya scene that also uses Caribou
- The scene open triggers a refresh of the panel, but these (presumably dying) widgets are still receiving events

johnhaddon commented:
How confident are you that you're fixing a bug in Gaffer here? The symptoms seem to imply that the real problem may be that Caribou is allowing a `GafferUI.Widget()` to die prematurely, leaving its Qt contents parented into a non-Gaffer window.

andrewkaufman commented:
I’m not at all confident, but when I saw we do the same thing in Widget.__enter and Wdget.__leave it seemed reasonable to add the protection in the other places as well.

I can check later today, but I think the ScriptWindow is the only widget Caribou takes ownership of. Maybe we need to unparent it from the maya ui element just before the CaribouBridge is destroyed.

johnhaddon commented:
I'm not averse to merging as a last resort because I know what debugging Qt embedded in Maya is like, and I don't think there's too much downside. But I do think it'd be worth doing a bit more investigation first; if there's something to be learnt, it'd be better to learn it than to just mask the problem. From the symptoms, my guess would be that this might be happening :

- Caribou is holding a top-level GafferUI.Widget, which owns a QtWidget.
- Caribou inserts the GafferUI.Widget._qtWidget() into a Maya layout. The Maya layout now owns the QtWidget, and will delete it only when the layout is deleted.
- Caribou is dropping the reference to the GafferUI.Widget before Maya deletes the layout.
- The QtWidget gets an event, we try to find the owning GafferUI.Widget, but can't because it's gone already.

andrewkaufman commented:
> Caribou is holding a top-level GafferUI.Widget, which owns a QtWidget.

The only GafferUI widgets Caribou owns are a `TabbedContainer` and a dict of `ScriptWindows`, both stored on member variables during the panel `__init__`

> Caribou inserts the GafferUI.Widget._qtWidget() into a Maya layout. The Maya layout now owns the QtWidget, and will delete it only when the layout is deleted.

This is done for the `TabbedContainer` in panel `__init__`

> Caribou is dropping the reference to the GafferUI.Widget before Maya deletes the layout.

I've yet to see a point where Caribou drops the reference to the `TabbedContainer` in my repro. We never explicitly call `del` on the member variable, and the panel `__del__` is only called when I actually close the floating maya window (and not called when I open the new file).

It does however drop the references to the `ScriptWindows` as each `ScriptNode` is removed from the app (which does happen on file->open). The `ScriptWindow` is removed from the member variable and from the `TabbedContainer` (via `removeChild`) and the weird reference-to-maya-parent-widgets for that `ScriptWindow` are removed as well.

> The QtWidget gets an event, we try to find the owning GafferUI.Widget, but can't because it's gone already.

Surely this is happening, but I don't know why we can't find the owning `GafferUI.Widget`. I'm guessing its the `ScriptWindow` we just dropped (since one of the errors is in Window.py eventFilter), but Maya shouldn't have ever thought it owned that QtWidget. It only owns the `TabbedContainer` one, which we still hold a reference to...

andrewkaufman commented:
Actually, I started addressing your comment about holding too many of those wonky maya widget references, and that seems to have solved the problem without needing these changes. So I guess we can close this one.

johnhaddon commented:
I just realised that the code for storing the wonky maya widget references is also storing a reference to the `ScriptWindow._qtWidget()`. In theory, we don't want to store that, we only want to store the Maya ones above it. It could be that storing that was the problem...

andrewkaufman commented:
Yeah I was wondering why it started with itself, but I was confident to change it.
